### PR TITLE
[SPARK-15145][ML]:port binary classification evaluator to spark.ml

### PR DESCRIPTION
**Pull request by wangmiao1981**
## What changes were proposed in this pull request?

Add accuracy into binary classification metrics.
## How was this patch tested?

(Please explain how this patch was tested. E.g. unit tests, integration tests, manual tests)
Add new scala unit test and all tests passed

### CODE REVIEW

File: mllib/src/main/scala/org/apache/spark/mllib/evaluation/binary/BinaryClassificationMetricComputers.scala [hunk 1]
```diff
@@ -24,6 +24,19 @@ private[evaluation] trait BinaryClassificationMetricComputer extends Serializabl
   def apply(c: BinaryConfusionMatrix): Double
 }
 
+/** Accuracy. Defined as 1.0 where there are zero no positive and negative examples. */
+private[evaluation] object Accuracy extends BinaryClassificationMetricComputer {
+  override  def apply(c: BinaryConfusionMatrix): Double = {
```
Comment (srowen) [inline mllib/src/main/scala/org/apache/spark/mllib/evaluation/binary/BinaryClassificationMetricComputers.scala]: Nit: extra space after override. You could also pull totalTrue into the else block.

File: mllib/src/test/scala/org/apache/spark/mllib/evaluation/BinaryClassificationMetricsSuite.scala [hunk 2]
```diff
@@ -80,8 +85,9 @@ class BinaryClassificationMetricsSuite extends SparkFunSuite with MLlibTestSpark
     val prCurve = Seq((0.0, 1.0)) ++ pr
     val f1 = pr.map { case (r, p) => 2.0 * (p * r) / (p + r)}
     val f2 = pr.map { case (r, p) => 5.0 * (p * r) / (4.0 * p + r)}
+    val accuracies = numTruePosNeg.map (t => t.toDouble / numAllCounts)
```
Comment (srowen) [inline mllib/src/test/scala/org/apache/spark/mllib/evaluation/BinaryClassificationMetricsSuite.scala]: Nit: space after map. It might fail style checks

**Comment by srowen** [comment]
I'm OK with it, as it's a somewhat common metric to want to see. Ideally, @mengxr would give a thumbs up/down as it's a new API method.

File: mllib/src/main/scala/org/apache/spark/mllib/evaluation/binary/BinaryClassificationMetricComputers.scala [hunk 3]
```diff
@@ -24,6 +24,19 @@ private[evaluation] trait BinaryClassificationMetricComputer extends Serializabl
   def apply(c: BinaryConfusionMatrix): Double
 }
 
+/** Accuracy. Defined as 1.0 where there are zero no positive and negative examples. */
```
Comment (MLnick) [inline mllib/src/main/scala/org/apache/spark/mllib/evaluation/binary/BinaryClassificationMetricComputers.scala]: "... are zero no positive and negative ..." should probably be "... are no positive or negative ..."?

**Comment by MLnick** [comment]
Should we not add `accuracy` to `ml` BinaryClassificationEvaluator too? That was my point really, the new method in mllib is more for usage within ml's evaluator.

**Comment by wangmiao1981** [comment]
@MLnick We can add accuracy to BinaryClassificationEvaluator. But we need to add new API of calculate accuracy as a Double. Now, it is RDD[Double, Double].

**Comment by srowen** [comment]
Yeah given how `BinaryClassificationEvaluator` works now, I don't think we could add accuracy, as it depends on picking a threshold for the classifier. Right now it only works with "global" metrics like AUC. @MLnick am I missing something there ... or is the conclusion maybe that that is another issue, allowing evaluation of a classifier that has already fixed a threshold?

**Comment by MLnick** [comment]
Ah right, ok. Good point. I was thinking more along the lines of adding `accuracy` to `BinaryClassificationEvaluator`. I don't think this makes sense here on a per threshold basis (it's like an "accuracy curve" which I can't say I've come across before). I'd prefer not to add this as a public API if it's not a well-used and defined concept. Sorry for the confusion @wangmiao1981 

I think we can make the JIRA about adding accuracy to the ml binary classification evaluator, and that may require the evaluator to compute the metric itself, not using `BinaryClassificationMetric` (it should be a simple computation).

**Comment by wangmiao1981** [comment]
@MLnick Do you want me to do "adding accuracy to the ml binary classification evaluator" in this JIRA or in a separate JIRA? Thanks!

**Comment by srowen** [comment]
Hm, yeah I also agree that I have not seen an 'accuracy curve'. Maybe we don't add that for now after all.

Nick how does the evaluator then work with respect to picking a threshold?

**Comment by MLnick** [comment]
The evaluator currently supports area under roc and pr. Those metrics are
across varying threshold levels. I don't see that they can actually be used
to pick a threshold, unless I'm missing something.

Accuracy could just be computed from the raw predicted and actual values as
#correct / total, but I guess that would be at the default threshold.
Unless the threshold is a param (which offhand I don't think it is), in
which case it could be cross-validated using, say, accuracy as a metric.

**Comment by srowen** [comment]
Yeah that's right, AUC gives a metric and it doesn't require a threshold. Yes using accuracy as a metric means somehow fixing a threshold. Just using "whatever's the default" is fairly reasonable. It bears some documentation. I wasn't sure if there was a reason that no other such metric was already implemented, like precision or recall, for this purpose.

**Comment by mengxr** [comment]
I wonder whether we should merge this into Spark 2.0 since we already passed the feature freeze deadline and the only pending items are linear algebra issues.

If we want to merge it into Spark 2.1, this conflicts with our decision on switching `spark.mllib` package into maintenance mode after 2.0. A better approach seems to be porting over existing binary classification metrics to `spark.ml` and adding this feature to it.

**Comment by srowen** [comment]
Agree, let's close this particular PR

**Comment by MLnick** [comment]
Yeah - Miao we can keep the JIRA open to look at accuracy in ml binary
evaluator. Probably for 2.1 now

**Comment by wangmiao1981** [comment]
@mengxr Thanks for your comments. As this one is a good-to-have feature, I agree to postpone it to 2.1 and I will change the JIRA to porting binary classification metrics to spark.ml. I can incorporate suggestions as we discussed above.

**Pull request closed without merging**